File: mosaic-tty/src/jvmMain/java/com/jakewharton/mosaic/tty/TtyInputStream.java
package com.jakewharton.mosaic.tty;

import java.io.IOException;
import java.io.InputStream;

final class TtyInputStream extends InputStream {
	private final Tty tty;

	TtyInputStream(Tty tty) {
		this.tty = tty;
	}

	@Override
	public int read() throws IOException {
		byte[] b = new byte[1];

		int read;
		do {
			read = read(b, 0, 1);
		} while (read == 0);

		return read == -1
			? -1
			: b[0] & 0xff;
	}

	@Override
	public int read(byte[] b, int off, int len) throws IOException {
		return tty.read(b, off, len);
	}
}

File: mosaic-tty/src/jvmMain/kotlin/com/jakewharton/mosaic/tty/TtyInputStream.kt
package com.jakewharton.mosaic.tty

import java.io.IOException
import java.io.InputStream

internal class TtyInputStream(
	private val tty: Tty,
) : InputStream() {
	private var closed = false

	override fun read(): Int {
		val buffer = ByteArray(1)
		while (true) {
			val read = read(buffer, 0, 1)
			if (read == -1) return -1
			if (read == 1) return buffer[0].toInt()
		}
	}

	override fun read(b: ByteArray, off: Int, len: Int): Int {
		if (!closed) {
			return tty.readInput(b, off, len)
		}
		throw IOException("closed")
	}

	override fun close() {
		closed = true
	}
}
